mapie/calibration.py

```diff
@@ -0,0 +1,559 @@
+from __future__ import annotations
+
+import warnings
+from typing import Dict, Optional, Tuple, Union, cast
+
+import numpy as np
+from sklearn.base import BaseEstimator, ClassifierMixin, RegressorMixin, clone
+from sklearn.calibration import _SigmoidCalibration
+from sklearn.isotonic import IsotonicRegression
+from sklearn.utils import check_random_state
+from sklearn.utils.multiclass import type_of_target
+from sklearn.utils.validation import (_check_y, _num_samples, check_is_fitted,
+                                      indexable)
+
+from ._typing import ArrayLike, NDArray
+from .utils import (check_cv, check_estimator_classification,
+                    check_estimator_fit_predict, check_n_features_in,
+                    check_null_weight, fit_estimator, get_calib_set)
+
+
+class MapieCalibrator(BaseEstimator, ClassifierMixin):
+    """
+    Calibration for multi-class problems.
+
+    This class performs calibration for various methods, currently only
+    top-label calibration [1].
+
+    Parameters
+    ----------
+    estimator : Optional[ClassifierMixin]
+        Any classifier with scikit-learn API
+        (i.e. with fit, predict, and predict_proba methods), by default
+        ``None``.
+        If ``None``, estimator defaults to a ``LogisticRegression`` instance.
+
+    method: Optional[str]
+        Method to choose for prediction interval estimates.
+        Choose among:
+
+        - "top_label", performs a calibration procedure on the class with
+           highest score, see section 2 of [1].
+
+        By default "top_label".
+
+    calibrator : Optional[Union[str, RegressorMixin]]
+        Any calibrator with scikit-learn API
+        (i.e. with fit, predict, and predict_proba methods), by default
+        ``None``.
+        If ``None``, calibrator defaults to a string ``sigmoid`` instance.
+
+        By default ``None``.
+
+    cv: Optional[str]
+        The cross-validation strategy for computing scores :
+
+        - ``split``, performs a standard splitting procedure into a
+          calibration and test set.
+        - ``"prefit"``, assumes that ``estimator`` has been fitted already.
+          All data provided in the ``fit`` method is then used
+          to calibrate the predictions through the score computation.
+          At prediction time, quantiles of these scores are used to estimate
+          prediction sets.
+
+        By default ``split``.
+
+    Attributes
+    ----------
+    valid_methods: List[str]
+        List of all valid methods.
+
+    classes_: NDArray
+        Array with the name of each class.
+
+    n_classes_: int
+        Number of classes that are in the training dataset.
+
+    uncalib_pred: NDArray
+        Array of the uncalibrated predictions.
+
+    calibrators: Dict[str, RegressorMixin]
+        Dictionnary of all the fitted calibrators.
```

Comment:
It is unclear for me whether calibrators are calibratred models or method to calibrate the model

Reply:
As referred to in the first sentence of the abstract of the paper for Top-Label calibration:
`A multiclass classifier is said to be top-label calibrated if the reported probability for the predicted
class—the top-label—is calibrated, conditioned on the top-label.`.

I think this would mean that they are a method to calibrate the output of the model.